Add Altera HAL support for Altera NIOSII MAX 10 devices

ramakrishnapallala opened the pull request.

nashif reviewed (changes requested):
please follow the process for importing external components as described here and adapt the commit message:

https://www.zephyrproject.org/doc/contribute/contribute_guidelines.html#identifying-contribution-origin

and https://www.zephyrproject.org/doc/contribute/contribute_non-apache.html

andrewboie commented on ext/hal/altera/altera_hal/HAL/inc/altera_common.h:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2017, Intel Corporation
```
Why are we doing full license boilerplate here, instead of doing SPDX-License-Identifier?

nashif replied:
this is not an original Zephyr file, we need to keep the original license boilerplate.

ramakrishnapallala replied:
anas, altera_common.h is added by me for Zephyr adoption of Altera HAL.

But do we need to keep SPDX or Intel license? I thought it's in external folder and thought Intel license header would be good.

ramakrishnapallala replied:
Andrew, the new ones are working fine. I can remove the old ones in reworked patches.

tyrantlizard77 reviewed (changes requested):
Do the new UART and TIMER drivers work on the existing CPU design?
I am still uneasy about keeping both the new and old drivers in tree.

nashif commented:
> anas, altera_common.h is added by me for Zephyr adoption of Altera HAL.

then this file needs to be APl licensed and marked as such with SPDX tagging

nashif commented:
I still do not see this following http://docs.zephyrproject.org/contribute/contribute_non-apache.html, where is the README?

ramakrishnapallala commented:
@nashif , i have added the README file. Please have a look.

nashif commented on ext/hal/altera/README:
```diff
@@ -0,0 +1,11 @@
+The sources in this directory are imported from Altera
```
Please follow the template provided in doc/contribute/code_component_README

ramakrishnapallala replied:
@nashif , i have actually provided info like Origin/Purpose/URL etc.. in the initial commit message itself. Also i have referred README's of other ext/hal/ folders are none of them are having this info README.
The README are simply having info like how they can be downloaded and what they do.

If you want, i can add that info in README as well.

nashif commented on ext/hal/altera/README:
```diff
@@ -0,0 +1,33 @@
+
+   The NIOSII port of Zephyr will be accessing the Altera HAL through
+   shim drivers to enable the different soft IP blocks on Zephyr.
+
```
You are missing the license here

nashif replied:
License:
   [use SPDX identifier (https://spdx.org/licenses/), such as BSD-3-Clause]

License Link:
   [link to the project's LICENSE file, e.g., https://github.com/xyz/LICENSE]

ramakrishnapallala replied:
for License, can i copy the whole License statement here?

also for License link, i can just put the same link as SDK download link, is that OK?

Commit:
ext: hal: altera: Add altera HAL sources in Zephyr

Add the Altera HAL sources from Altera Quartus Lite SDK
to Zephyr OS under 'ext/hal/altera' folder.

Origin: Altera Quartus Prime Lite SDK
URL: http://dl.altera.com/?edition=lite
Version: v17.0
Purpose: Provides HAL support to Altera's soft IP's
Maintained-by: External

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
ext: hal: altera: Add Altera HAL into Zephyr build system

Make Altera HAL drivers compilable by adding them into
Zephyr build system.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
ext: hal: altera: Add basic infrastructure to enable HAL drivers

Add the basic infrastructure like driver <-> IRQ mapping table
and ZEPHYR_RTOS compile time flag which are neccessary to make
use the HAL drivers from Zephyr shim drivers.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
drivers: timer: Add Altera shim driver for timer soft IP

Delete the native timer soft IP driver as we will be reusing
the Altera's HAL drivers for most of the soft IP's.

Add shim driver support for Altera timer system clock soft IP.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
drivers: serial: Add Altera shim driver for JTAG UART soft IP

Delete the native UART JTAG driver as we will be reusing
the Altera's HAL driver.

Add the shim driver support for Altera HAL's JTAG UART soft IP.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
ext: hal: altera: Add Altera HAL README file

Add Altera HAL README file as per Zephyr Guidelines.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

nashif commented:
recheck

nashif commented:
the external components in this PR have been approved with majority on both TSC&BOARD, merging for 1.10, action will be taken if there will be any objections.

ramakrishnapallala commented:
Ok thanks @nashif  :)
